aspnetcore/blazor/javascript-interoperability/import-export-interop.md

```diff
@@ -0,0 +1,417 @@
+    public partial class CallJavaScript1
+    {
+        [JSImport("getMessage", "CallJavaScript1")]
+        internal static partial string GetWelcomeMessage();
+    }
+}
+```
+
+The app's namespace for the preceding `CallJavaScript1` partial class is `BlazorServer`. The component's namespace is `BlazorSample.Pages`. If using the preceding component in a local test app, update the namespace to match the app. For example, the namespace is `ContosoApp.Pages` if the app's namespace is `ContosoApp`. For more information, see <xref:blazor/components/index#partial-class-support>.
+
+In the imported method signature, you can use .NET types for parameters and return values, which are marshalled automatically by the runtime. Use `JSMarshalAsAttribute<T>` to control how the imported method parameters are marshalled. For example, you might choose to marshal a `long` as <xref:System.Runtime.InteropServices.JavaScript.JSType.Number?displayProperty=nameWithType> or <xref:System.Runtime.InteropServices.JavaScript.JSType.BigInt?displayProperty=nameWithType>. You can pass <xref:System.Action>/<xref:System.Func%601> callbacks as parameters, which are marshalled as callable JS functions. You can pass both JS and managed object references, and they are marshaled as proxy objects, keeping the object alive across the boundary until the proxy is garbage collected. You can also import and export asynchronous methods with a <xref:System.Threading.Tasks.Task> result, which are marshaled as [JS promises](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Promise). Most of the marshalled types work in both directions, as parameters and as return values, on both imported and exported methods, which are covered in the [Call .NET from JavaScript](#call-net-from-javascript) section later in this article.
+
+The module name in the `[JSImport]` attribute and the call to load the module in the component, which is covered next with `JSHost.ImportAsync` syntax, should match. The module name must also be unique in the app. When authoring a library for deployment in a NuGet package, we recommend using the NuGet package namespace as a prefix in module names. In the following example, the module name reflects the `Contoso.InteropServices.JavaScript` package and a folder of user message interop classes (`UserMessages`):
+
+```csharp
+[JSImport("getMessage", 
+    "Contoso.InteropServices.JavaScript.UserMessages.CallJavaScript1")]
+```
+
+If the JS function doesn't directly interact with the rendered Document Object Model (DOM), import the module in [`OnInitializedAsync`](xref:blazor/components/lifecycle#component-initialization-oninitializedasync). Call the imported JS function with the .NET interop method.
+
+If the JS module interacts directly with the component's rendered UI, import the module in [`OnAfterRenderAsync`](xref:blazor/components/lifecycle#after-component-render-onafterrenderasync) and call the imported JS function:
```

Comment:
hummmm 🤔 ... I think Steve implied that if ... for example ... some JS lib is imported **_by the library being used_** (perhaps from the Internet), that you wouldn't be guaranteed that its there that early. However, that's just a guess. I think we'd need to ask him about it.

Yes, tho ... I see what you mean here in this case ... in this particular example.

Reply:
No ... actually ... I think it might be this in the module of Steve's example ...

```javascript
mapboxgl.accessToken = '{ACCESS TOKEN}';
```

Doesn't the module need to load after render for that `accessToken` set to occur?

Reply:
> Doesn't the module need to load after render for that accessToken set to occur?

I'm not an expert on ES6 modules, but I think the code in the module doesn't run until all of the modules it `import`s are loaded (so the `mapboxgl` object should be available by the time that line runs).

But either way, I don't think the component getting rendered would have any impact on the mechanics of the import procedure (with a possible exception being that the module initialization somehow requires the DOM to look a certain way, but this would be highly unusual).

The `addMapToElement()` invocation in Steve's example _does_ have to be called after the component is rendered, because it utilizes the DOM rendered by the component. I'm guessing the module import was also placed in `OnAfterRenderAsync()` because JS interop in `OnInitializedAsync()` won't work during Blazor Server prerendering. But that isn't a concern in the context of this new JS interop module, since it only works in WebAssembly anyway.

cc @SteveSandersonMS in case there's something I'm missing!

Reply:
:+1: ... I think the changes that I made on the commit resolve this aspect then. The only reference made to after-render now is for JS interop that mutates the DOM with the warning about the special cases when that's required (e.g., 3rd party lib).